test/drbgtest.c

```diff
@@ -724,31 +739,222 @@ static int test_drbg_reseed(int expect_success,
+        close(fd[0]);
+
+        /* check whether all three DRBGs reseed and send output to parent */
+        if (TEST_true(test_drbg_reseed(1, master, public, private,
+                                        &random[0], &random[RANDOM_SIZE],
+                                       1, 1, 1, 0))
+            && TEST_true(write(fd[1], random, sizeof(random))
+                         == sizeof(random))) {
+
+            rv = 1;
+        }
+
+        /* close the write end */
+        close(fd[1]);
+
+        /* convert boolean to exit code */
+        exit(rv == 0);
     }
+}
+
+static int test_rand_drbg_reseed_on_fork(RAND_DRBG *master,
+                                         RAND_DRBG *public,
+                                         RAND_DRBG *private)
+
+{
+    int i;
+    pid_t pid = getpid();
+
+    int duplicate = 0;
+
+    unsigned char random[2*RANDOM_SIZE];
+
+
+    drbg_fork_result result[DRBG_FORK_RESULT_COUNT];
+    drbg_fork_result sorted[DRBG_FORK_RESULT_COUNT];
+    drbg_fork_result *presult = &result[2];
+
+    memset(&result,  0, sizeof(result));
+
+    for (i = 1 ; i <= DRBG_FORK_COUNT ; ++i) {
+
+        presult[0].pindex = presult[1].pindex = i;
+
+        sprintf(presult[0].name, "child %d", i);
+        strcpy(presult[1].name, presult[0].name);
 
-    /* I'm the child; check whether all three DRBGs reseed. */
-    if (!TEST_true(test_drbg_reseed(1, master, public, private, 1, 1, 1, 0)))
-        status = 1;
-    exit(status);
+        /* collect the random output of the children */
+        if (!TEST_true(test_drbg_reseed_in_child(master,
+                                                 public,
+                                                 private,
+                                                 presult)))
+            return 0;
+
+        presult += 2;
+    }
+
+    /* collect the random output of the parent */
+
+    if (!TEST_true(test_drbg_reseed(1,
+                                    master, public, private,
+                                    &random[0], &random[RANDOM_SIZE],
+                                    0, 0, 0, 0)))
+        return 0;
+
+    strcpy(result[0].name, "parent");
+    strcpy(result[1].name, "parent");
+
+    /* output of public drbg */
+    result[0].pid = pid;
+    result[0].private = 0;
+    memcpy(result[0].random, &random[0], RANDOM_SIZE);
+
+    /* output of private drbg */
+    result[1].pid = pid;
+    result[1].private = 1;
+    memcpy(result[1].random, &random[RANDOM_SIZE], RANDOM_SIZE);
+
+    /* sort the entries... */
+    memcpy(sorted, result, sizeof(sorted));
+    qsort(sorted, DRBG_FORK_RESULT_COUNT, sizeof(drbg_fork_result),
+          compare_drbg_fork_result);
+
+    /* ...and search for duplicates in the first random byte */
+    for (i=1 ; i < DRBG_FORK_RESULT_COUNT ; ++i) {
+        if (sorted[i].random[0] == sorted[i-1].random[0]) {
+            ++duplicate;
+        }
+    }
+
+    if (duplicate > DRBG_FORK_COUNT) {
```

Comment:
So you're taking 16 readings (8 * 2), and considering the test failed if any 9 of them are the same byte (i.e 1/256). This sounds like the birthday problem from probability theory...I wonder if the chances of this happening are higher than we might expect? The maths is beyond me though. Are we over complicating this? I wonder if a simple test which just checks for a single duplicate among any of the outputs is sufficient (although that too is the birthday problem)

Reply:
It's a binomial distribution with p = 1/256 and n=16.  The cumulative distribution is related to the incomplete beta function.  An online [high precision calculator](https://keisan.casio.com/exec/system/1180573199), gives the probability at around 2.4e-18.  [Wolfram alpha](https://www.wolframalpha.com/input/?i=binomial+distribution+n%3D16%2C+p%3D1%2F256%2C+x%3E%3D9) concurs.

Reply:
The Wolfram alpha link is the easier.

Reply:
Note to self: don't do probability while tired with a headache.  Of course I got it wrong.  The probability will be much higher: **1.7e-13**.

``` R
p <- 1/256;
i <- 1:8;
# pbinom is Pr(X<=x) and but want Pr(X>=x), hence the 1 - and 7 instead of 8:
sum((1-p)^(i-1) * (1 - pbinom(7, 16-i, p)))   
# This is subject to severe cancellation, so flip the binomial around:
sum((1-p)^(i-1) * pbinom(16-i-7, 16-i, 1-p))
```